src/udp.rs

```diff
@@ -349,142 +289,133 @@ impl Inner {
         }
     }
 
-    fn write_to<T: Serialize>(&mut self, msg: Option<(T, SocketAddr, Priority)>) -> ::Res<bool> {
-        let expired_keys: Vec<u8> = self
-            .write_queue_2
-            .iter()
-            .skip_while(|&(&priority, queue)| {
-                priority < self.conf.msg_drop_priority || // Don't drop high-priority messages.
-                queue.front().map_or(true, |&(ref timestamp, _, _)| {
-                    timestamp.elapsed().as_secs() <= self.conf.max_msg_age_secs
-                })
-            }).map(|(&priority, _)| priority)
-            .collect();
-        let dropped_msgs: usize = expired_keys
-            .iter()
-            .filter_map(|priority| self.write_queue_2.remove(priority))
-            .map(|queue| queue.len())
-            .sum();
-        if dropped_msgs > 0 {
-            trace!(
-                "Insufficient bandwidth. Dropping {} messages with priority >= {}.",
-                dropped_msgs,
-                expired_keys[0]
-            );
-        }
+    /// Returns `Ok(false)`, if write to the underlying stream would block.
+    fn flush_write_to_until_would_block(&mut self) -> ::Res<bool> {
+        loop {
+            let (data, peer) = if let Some(data_peer) = self
+                .current_write_2
+                .take()
+                .or_else(|| self.out_queue2.next_msg())
+            {
+                data_peer
+            } else {
+                return Ok(true);
+            };
 
-        if let Some((msg, peer, priority)) = msg {
-            let entry = self
-                .write_queue_2
-                .entry(priority)
-                .or_insert_with(|| VecDeque::with_capacity(10));
-            entry.push_back((Instant::now(), serialise(&msg)?, peer));
+            let res = self.sock.send_to(&data, &peer);
+            handle_send_res(res, &mut self.current_write_2, data.len(), (data, peer))?;
         }
+    }
+}
 
-        loop {
-            let (data, peer) = if let Some(pkg) = self.current_write_2.take() {
-                pkg
-            } else {
-                let (key, (_time_stamp, data, peer), empty) =
-                    match self.write_queue_2.iter_mut().next() {
-                        Some((key, queue)) => (*key, unwrap!(queue.pop_front()), queue.is_empty()),
-                        None => return Ok(true),
+/// Helper trait for `handle_recv_res()`.
+trait IsEmpty {
+    fn is_empty(&self) -> bool;
+}
+
+impl IsEmpty for Vec<u8> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl IsEmpty for (Vec<u8>, SocketAddr) {
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+fn handle_recv_res<T: IsEmpty>(
+    recv_res: io::Result<T>,
+    read_buffer: &mut VecDeque<T>,
+    would_block: &mut bool,
+) -> ::Res<Option<T>> {
+    match recv_res {
+        Ok(msg) => {
+            if msg.is_empty() {
+                let e = Err(SocketError::ZeroByteRead);
```

Comment:
Ok this is a little more involved. Check [point 1 here](https://github.com/ustulation/p2p/commit/129c65623bc730061c45150a0316e49744e3dd38). 0B read is a valid read in UDP if you sent a 0 byte datagram. However on windows it some failures to a connected UDP socket cause spurious 0B reads. So the correct way to implement this would be to ignore the 0B read (as it could denote either a failure or a successful read of a genuine 0B datagram in which case you should do the usual reading until `wouldblock`), push nothing into the read-buffer and indicate the caller code (`read<>` or `read_frm<>`) that they should continue reading (so as if something was read successfully).

So this is where it'll differ from the TCP's implementation of reading raw data from wire (where a 0B read would mean a genuine EOF)